### PR TITLE
trans: Only instantiate #[inline] functions in codegen units referencing them

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR changes how `#[inline]` functions are translated. Before, there was one "master instance" of the function with `external` linkage and a number of on-demand instances with `available_externally` linkage in each codegen unit that referenced the function. This had two downsides:
- Public functions marked with `#[inline]` would be present in machine code of libraries unnecessarily (see #36280 for an example)
- LLVM would crash on `i686-pc-windows-msvc` due to what I suspect to be a bug in LLVM's Win32 exception handling code, because it doesn't like `available_externally` there (#36309).

This PR changes the behavior, so that there is no master instance and only on-demand instances with `internal` linkage. The downside of this is potential code-bloat if LLVM does not completely inline away the `internal` instances because then there'd be N instances of the function instead of 1. However, this can only become a problem when using more than one codegen unit per crate.

cc @rust-lang/compiler

### CODE REVIEW

**Comment by retep998** [comment]
I just tested this PR locally, and it appears to work mostly. According to `-Ztime-passes` llvm passes has gone down by over an order of magnitude for winapi. Translation went down also, but not as significantly. The LLVM IR went down from 2,169KiB to 99KiB, most of the remainder is simply debuginfo.

However, it is still generating code for methods. Here's a quote from `--pretty-expanded`. Note the `#[inline]`.
```Rust
        impl IMalloc {
            #[inline]
            pub unsafe fn Alloc(&mut self, cb: SIZE_T) -> *mut c_void {
                ((*self.lpVtbl).Alloc)(self, cb)
            }
```
And here it is still emitting LLVM IR.
```ll
; Function Attrs: uwtable
define %"ctypes::c_void"* @_ZN6winapi2um10objidlbase7IMalloc7Realloc17h12cc9c5a9dd2c848E(%"um::objidlbase::IMalloc"* dereferenceable(8), %"ctypes::c_void"*, i64) unnamed_addr #0 {
entry-block:
  %3 = getelementptr inbounds %"um::objidlbase::IMalloc", %"um::objidlbase::IMalloc"* %0, i64 0, i32 0
  %4 = load %"um::objidlbase::IMallocVtbl"*, %"um::objidlbase::IMallocVtbl"** %3, align 8
  %5 = getelementptr inbounds %"um::objidlbase::IMallocVtbl", %"um::objidlbase::IMallocVtbl"* %4, i64 0, i32 2
  %6 = load %"ctypes::c_void"* (%"um::objidlbase::IMalloc"*, %"ctypes::c_void"*, i64)*, %"ctypes::c_void"* (%"um::objidlbase::IMalloc"*, %"ctypes::c_void"*, i64)** %5, align 8
  %7 = tail call %"ctypes::c_void"* %6(%"um::objidlbase::IMalloc"* nonnull %0, %"ctypes::c_void"* %1, i64 %2)
  ret %"ctypes::c_void"* %7
}
```

**Comment by pcwalton** [comment]
@retep998 Wow! 😲

**Comment by michaelwoerister** [comment]
@retep998 Phew! And I was already wondering why I couldn't reproduce your results `:)`

It's to be expected that translation still takes some time, since that also includes the time for exporting metadata.

File: src/librustc_trans/trans_item.rs [hunk 1]
```diff
@@ -282,6 +273,18 @@ impl<'a, 'tcx> TransItem<'tcx> {
...
+    /// Returns true if there has to be a local copy of this TransItem in every
+    /// codegen unit that references it (as with inlined functions, for example)
+    pub fn needs_local_copy(&self, tcx: TyCtxt<'a, 'tcx, 'tcx>) -> bool {
+        // Currently everything that is instantiated only on demand is done so
+        // with "internal" linkage, so we need a copy to be present in every
+        // codegen unit.
+        // This is coincidental: We could also instantiate something only if it
+        // is referenced (e.g. a regular, private function) but place it in its
+        // own codegen unit with "external" linkage.
+        self.is_instantiated_only_on_demand(tcx)
```
Comment (nikomatsakis) [inline src/librustc_trans/trans_item.rs]: Nit: it might be nice to put a few notes here as to *why* things `need_local_copy`

**Comment by michaelwoerister** [comment]
@bors retry

Spurious? cc @alexcrichton

**Pull request merged**